plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go

```diff
@@ -250,89 +285,204 @@ func (c *CiscoTelemetryMDT) handleTelemetry(data []byte) {
+	case *telemetry.TelemetryField_DoubleValue:
+		return val.DoubleValue
+	case *telemetry.TelemetryField_FloatValue:
+		return val.FloatValue
 	}
-	namebuf.WriteString(strings.Replace(field.Name, "-", "_", -1))
+	return nil
+}
 
-	// Decode Telemetry field value if set
-	var value interface{}
+func decodeTag(field *telemetry.TelemetryField) string {
 	switch val := field.ValueByType.(type) {
 	case *telemetry.TelemetryField_BytesValue:
-		value = val.BytesValue
+		return string(val.BytesValue)
 	case *telemetry.TelemetryField_StringValue:
-		value = val.StringValue
+		return val.StringValue
 	case *telemetry.TelemetryField_BoolValue:
-		value = val.BoolValue
+		if val.BoolValue {
+			return "true"
+		}
+		return "false"
 	case *telemetry.TelemetryField_Uint32Value:
-		value = val.Uint32Value
+		return strconv.FormatUint(uint64(val.Uint32Value), 10)
 	case *telemetry.TelemetryField_Uint64Value:
-		value = val.Uint64Value
+		return strconv.FormatUint(val.Uint64Value, 10)
 	case *telemetry.TelemetryField_Sint32Value:
-		value = val.Sint32Value
+		return strconv.FormatInt(int64(val.Sint32Value), 10)
 	case *telemetry.TelemetryField_Sint64Value:
-		value = val.Sint64Value
+		return strconv.FormatInt(val.Sint64Value, 10)
 	case *telemetry.TelemetryField_DoubleValue:
-		value = val.DoubleValue
+		return strconv.FormatFloat(val.DoubleValue, 'f', -1, 64)
 	case *telemetry.TelemetryField_FloatValue:
-		value = val.FloatValue
+		return strconv.FormatFloat(float64(val.FloatValue), 'f', -1, 32)
+	default:
+		return ""
+	}
+}
+
+// Recursively parse tag fields
+func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemetry.TelemetryField, prefix string) {
+	localname := strings.Replace(field.Name, "-", "_", -1)
+	name := localname
+	if len(name) == 0 {
+		name = prefix
+	} else if len(prefix) > 0 {
+		name = prefix + "/" + localname
 	}
 
-	if value != nil {
-		// Distinguish between tags (keys) and fields (data) to write to
-		if fields != nil {
-			fields[namebuf.String()] = value
+	if tag := decodeTag(field); len(tag) > 0 {
+		localname := strings.Replace(field.Name, "-", "_", -1)
+		if _, exists := tags[localname]; !exists { // Use short keys whenever possible
+			tags[localname] = tag
 		} else {
-			if _, exists := tags[field.Name]; !exists { // Use short keys whenever possible
-				tags[field.Name] = fmt.Sprint(value)
-			} else {
-				tags[namebuf.String()] = fmt.Sprint(value)
+			tags[name] = tag
+		}
+	}
+
+	for _, subfield := range field.Fields {
+		c.parseKeyField(tags, subfield, name)
+	}
+}
+
+func (c *CiscoTelemetryMDT) parseContentField(grouper *metric.SeriesGrouper, field *telemetry.TelemetryField, prefix string,
+	path string, tags map[string]string, timestamp time.Time) {
+	name := strings.Replace(field.Name, "-", "_", -1)
+	if len(name) == 0 {
+		name = prefix
+	} else if len(prefix) > 0 {
+		name = prefix + "/" + name
+	}
+
+	extraTags := c.extraTags[path+"/"+name]
+
+	if value := decodeValue(field); value != nil {
```

Comment:
This seems like an error, can we return early if decode fails?  We should try to do this whenever we can, if the message isn't valid then lets skip it instead of trying to produce something which may not be right.
```
value, err := decodeValue(field)
if err != nil
    return err
}
```

Reply:
value being nil is legitimate when the field has children. usually a field has either children or a value (e.g. a tree-structure where only leafs carry values). I have added a return statement inside this branch, i.e. if it is a value, stop processing children.